File: apps/web/src/types/me.ts
/**
 * `session-identity-api` wire shape (AUTH-07, frozen), consumed by
 * `meApi.ts::fetchMe()` (DECISIONS.md D-01, DATA-DESIGN.md § 9 Contract).
 */

/**
 * `GET /api/me`'s response body (AUTH-07) -- **exactly** these two fields.
 * The backend's response model sets `extra="forbid"`, so no `jobTitle` /
 * `email` / `groups` ever exists on the wire to read; `jobTitle` is composed
 * frontend-side only, via `composeSignedInUser.ts` (DECISIONS.md D-03,
 * research condition 1). `name` is `null` for a `/auth/dev-bypass` token,
 * which carries no profile claims (README.md `GET /api/me` row) -- never
 * composed from `email`. `persona` is the resolver's output verbatim, never
 * inferred by a consumer from its own route.
 */
export interface MeData {
  name: string | null;
  persona: string;
}

/**
 * `fetchMe()`'s result (DATA-DESIGN.md § 9 Contract). Mirrors
 * `OverviewSummaryResult`'s (`@/types/overview`) status vocabulary exactly --
 * `"forbidden"` (403, persona resolution failed) checked before
 * `"unauthorized"` (401) in `meApi.ts`, both distinct from the generic
 * `"error"` catch-all (network/timeout throw or any other non-ok status).
 */
export type MeResult =
  | { status: "ok"; data: MeData }
  | { status: "forbidden" }
  | { status: "unauthorized" }
  | { status: "error" };

File: services/api/app/schemas/me.py
"""Response schema for GET /api/me (AUTH-07-AC-1, `session-identity-api`, api.md).

`MeResponse` is the sealed `session-identity-api` envelope -- exactly `name` and
`persona`, nothing else. `extra="forbid"` makes the exact-two-field contract a
runtime guarantee, not just a convention: no additional field (e.g. `email`,
`groups`, `programs`, `jobTitle`) can ever leak onto the wire by accident.
"""

from pydantic import BaseModel, ConfigDict, Field


class MeResponse(BaseModel):
    """Session identity: display name + resolved persona (AUTH-07-AC-1).

    `name` is the verified token's OIDC profile-claim fallback chain
    (`CurrentUser.name`, AC-2) -- `None` when no profile claim is present
    (e.g. a `/auth/dev-bypass` token), never a fabricated placeholder. `persona`
    is `persona-resolver.resolve(role)`'s output verbatim (AC-3) -- never
    hardcoded by the route and never defaulted; a resolution failure raises
    before a response model is ever constructed (AC-4).
    """

    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(..., description="Display name, or None with no profile claims")
    persona: str = Field(..., description="persona-resolver.resolve(role)'s output, verbatim")
